uid/src/main/java/com/tomtom/kotlin/uid/Uid.kt

```diff
@@ -0,0 +1,217 @@
+ *
+ * @param T Type tag for ID, to make IDs type-safe.
+ */
+class Uid<T> {
+    private val uuid: String
+
+    /**
+     * Create a new, unique UUID-based ID.
+     */
+    constructor() {
+        uuid = UUID.randomUUID().toString()
+    }
+
+    /**
+     * Instantiates a [Uid] with a string. Mainly used when de-serializing existing entities.
+     *
+     * The format of 'uuid' is checked to comply with a standard UUID format, which is:
+     * - Dashes at positions 8, 13, 18, 23 (base 0).
+     * - Characters 0-9 and a-f (lowercase only).
+     *
+     * If this format is used, the creation of the [Uid] is very fast. If an alternative format
+     * us used, as accepted by [fromString], the call is much more expensive.
+     *
+     * @param uuidAsString Existing string ID.
+     * @throws IllegalArgumentException If name does not conform to the string representation
+     * as described in [UUID.toString]. Use [isValid] to make sure the string is valid.
+     */
+    constructor(uuidAsString: String) {
+        /**
+         * This code has been optimized to NOT just call UUID.fromString(uuid) to convert the
+         * UUID-String into a String (and catch an IllegalArgumentException).
+         *
+         * If the UUID does not comply, the expensive call to UUID.fromString is made after all.
+         */
+        val length = uuidAsString.length
+        require(length in UUID_MIN_LENGTH..UUID_MAX_LENGTH) {
+            "Length of UUID must be [" + UUID_MIN_LENGTH + ", " +
+                UUID_MAX_LENGTH + "], but is " + uuidAsString.length + ", uuid=" + uuidAsString
+        }
+
+        // Check dashes.
+        this.uuid = if (areDashesAtCorrectPosition(uuidAsString)) {
+            uuidAsString.toLowerCase().also {
+                require(onlyContainsValidUuidCharacters(it)) {
+                    "Incorrect UUID format, uuid=$uuidAsString"
+                }
+            }
+        } else {
+            UUID.fromString(uuidAsString).toString().toLowerCase()
+        }
+    }
+
+    /**
+     * Instantiates an ID with a [UUID].
+     *
+     * @param uuidAsUuid Existing [UUID].
+     */
+    private constructor(uuidAsUuid: UUID) {
+        this.uuid = uuidAsUuid.toString()
+    }
+
+    /**
+     * Returns hex string representation of this Uid. Opposite of [fromHexString].
+     * This representation is shorthand, it does not include dashes.
+     *
+     * @return Hex string representation of ID, exactly 32 characters long.
+     */
+    fun toHexString(): String {
+        val uuid2 = UUID.fromString(uuid)
+        val msb = java.lang.Long.toHexString(uuid2.mostSignificantBits)
+        val lsb = java.lang.Long.toHexString(uuid2.leastSignificantBits)
+        val sb = StringBuilder()
+        for (i in 1..16 - msb.length) {
+            sb.append('0')
+        }
+        sb.append(msb)
+        for (i in 1..16 - lsb.length) {
+            sb.append('0')
+        }
+        sb.append(lsb)
+        val value = sb.toString()
```

Comment:
Or just: `val value = "${msb.padEnd(16)}${lsb.padEnd(16)}"`

But I'm not sure if that is more efficient.

Reply:
Should be padStart.